### PR TITLE
Upstream Hostname into operator-rs

### PR DESCRIPTION
**Pull request by nightkr**
# Description

It's becoming useful in other operators, so let's try to share it.

As a bonus, by reusing the validation machinery in op-rs we can also add the validation rules to the CRD.

Op-rs equivalent: https://github.com/stackabletech/operator-rs/pull/851

## Definition of Done Checklist

- Not all of these items are applicable to all PRs, the author should update this template to only leave the boxes in that are relevant
- Please make sure all these things are done and tick the boxes

```[tasklist]
# Author
- [x] Changes are OpenShift compatible
- [x] Helm chart can be installed and deployed operator works
- [x] Integration tests passed (for non trivial changes)
- [x] Changes need to be "offline" compatible
```

```[tasklist]
# Reviewer
- [ ] Code contains useful comments
- [ ] Code contains useful logging statements
- [ ] (Integration-)Test cases added
- [ ] Documentation added or updated. Follows the [style guide](https://docs.stackable.tech/home/nightly/contributor/docs/style-guide).
- [ ] Changelog updated
- [ ] Cargo.toml only contains references to git tags (not specific commits or branches)
```

```[tasklist]
# Acceptance
- [ ] Feature Tracker has been updated
- [ ] Proper release label has been added
- [ ] [Roadmap](https://github.com/orgs/stackabletech/projects/25/views/1) has been updated
```

### CODE REVIEW

**Comment by nightkr** [comment]
Tests running: https://testing.stackable.tech/view/02%20Operator%20Tests%20(custom)/job/secret-operator-it-custom/7/, https://testing.stackable.tech/view/02%20Operator%20Tests%20(custom)/job/secret-operator-it-custom/8/

**Comment by nightkr** [comment]
The tests are green against kind, but red against OpenShift due to some SCC profile errors.. @razvan could this be related to #488?

**Comment by razvan** [comment]
> The tests are green against kind, but red against OpenShift due to some SCC profile errors.. @razvan could this be related to #488?

No. It looks  like you forgot to pass `--test-suite openshift` to the job parameters.

**Comment by nightkr** [comment]
Oh, that's not automatic? Gotcha.

**Comment by nightkr** [comment]
OpenShift round 2: https://testing.stackable.tech/view/02%20Operator%20Tests%20(custom)/job/secret-operator-it-custom/9/

**Comment by nightkr** [comment]
Openshift is green once running the correct suite.. :P

**Review by NickLarsenNZ — APPROVED** [review]
LGTM

**Comment by sbernauer** [comment]
FYI, I put this issue on the board so that everyone has a chance to follow development :)

File: CHANGELOG.md [hunk 1]
```diff
@@ -9,6 +9,12 @@ All notable changes to this project will be documented in this file.
 - Active Directory's `samAccountName` generation can now be customized ([#454]).
 - Added experimental cert-manager backend ([#482]).
 
+### Changed
+
+- Refactored hostname validation ([#494]).
+  - BREAKING: Hostname validation is now somewhat stricter.
+  - BREAKING: Hostname validation is now enforced in CRD.
```
Comment (nightkr) [inline CHANGELOG.md:16]: @sbernauer This regressed in upstream op-rs (https://github.com/stackabletech/operator-rs/pull/863/files/87ca3050bda8070d36076706bbea41f2ee008e3d#diff-5ca013a250a561e7d4a42a3272200ba0dc20b0ad8345fb7660c904de3407f791).

Comment (sbernauer) [reply CHANGELOG.md:16]: Yep, we had some discussions around RFC 1123 and renamed it accordingly to the RFC (at least we hope ^^), before it gaines widespread adoption. We would have involved you but you where on vacation and we did not want to further delay the PR, but please feel free to veto, happy to discuss with @Techassi!

**Review by NickLarsenNZ — APPROVED** [review]
LGTM

**Pull request merged**